library/net.c

```diff
@@ -221,7 +258,41 @@ static int dnet_io_req_queue(struct dnet_net_state *st, struct dnet_io_req *orig
 	int err = 0;
 	struct dnet_io_req *r;
 
-	r = dnet_io_req_copy(st, orig);
+	/*
+	 * If destination is local node there is no need to copy it via net thread.
+	 * Here it could be sheduled to the proper IO pool directly.
+	 * If io_req hold fd instead of data it should be read here.
+	 */
+	if (st == st->n->st) {
+		dnet_log(st->n, DNET_LOG_DEBUG, "Sending to local state: data: 0x%lx data_size: %d", (unsigned long)orig->data, (int)orig->dsize);
+
+		r = dnet_io_req_copy(st, orig, 1);
+		if (!r) {
+			err = -ENOMEM;
+			goto err_out_exit;
+		}
+
+		/*
+		 * Fixup request.
+		 * r->header and r->data are just a pointers to the same buffer.
+		 * Here r->header could be any size, it depend on function that created io_req.
+		 * In net thread r->hsize is ALWAYS sizeof(struct dnet_cmd) and r->data points
+		 * exactly after struct dnet_cmd struct. We need to modify our copy to the same layout
+		 * because dnet_schedule_io requires it.
+		 */
+		if (r->hsize != sizeof(struct dnet_cmd)) {
```

Comment:
>For dnet_schedule_io header size must be sizeof(struct dnet_cmd). 

That's wrong assumption, header must be at least `sizeof(struct dnet_cmd)` since this is how elliptics protocol works.
Header size is a boundary which is needed when **sender** wants to send some preallocated data plus data from file descriptor, for instance sending read header (`struct dnet_cmd` + `struct dnet_io_attr`) + eblob content from file descriptor via `sendfile()`.

Receiver side only uses header size and pointer for informational purposes since receiver reads whole command content (all headers + data) into single continuous buffer.